feat: add `stats/incr/nanmmeanstdev`

saad-imran-dev opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #5591

## Description

> What is the purpose of this pull request?

This pull request:

-   It adds module which is a wrapper around `mmeanstdev` for handling `NaN` values. It ignores any `NaN` values passes in the accumulator

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5591

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

saad-imran-dev committed:
feat: add stats/incr/nanmmeanstdev

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

saad-imran-dev committed:
feat: add stats/incr/nanmmeanstdev

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

ShabiShett07 commented:
@saad-imran-dev, please update your pr title to:
```
feat: add `stats/incr/nanmmeanstdev`
```
to pass the pr title check.

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/package.json:
```diff
@@ -0,0 +1,80 @@
+      "url": "https://github.com/stdlib-js/stdlib/graphs/contributors"
+    }
+  ],
+  "main": "lib",
```
```suggestion
  "main": "./lib",
```

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/package.json:
```diff
@@ -0,0 +1,80 @@
+    "benchmark": "benchmark",
+    "doc": "docs",
+    "example": "examples",
+    "lib": "lib",
```
```suggestion
    "lib": "./lib",
```

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/package.json

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/package.json

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev committed:
Update package.json

Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/benchmark/benchmark.js:
```diff
@@ -0,0 +1,69 @@
+// MODULES //
+
+var bench = require( '@stdlib/bench' );
+var randu = require( '@stdlib/random/base/randu' );
```
```suggestion
```
I think we don't need to use this package, instead we could just pass iteration number as input

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/benchmark/benchmark.js:
```diff
@@ -0,0 +1,69 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		v = acc( randu() );
```
```suggestion
		v = acc( i );
```
Same reason

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md:
```diff
@@ -0,0 +1,234 @@
+
+#### incrnanmmeanstdev( \[out,] window )
+
+Returns an accumulator `function` which incrementally computes a moving [arithmetic mean][arithmetic-mean] and [corrected sample standard deviation][standard-deviation]. The `window` parameter defines the number of values over which to compute the moving [arithmetic mean][arithmetic-mean] and [corrected sample standard deviation][standard-deviation].
```
```suggestion
Returns an accumulator function which incrementally computes a moving [arithmetic mean][arithmetic-mean] and [corrected sample standard deviation][standard-deviation], ignoring `NaN` values. The `window` parameter defines the number of values over which to compute the moving [arithmetic mean][arithmetic-mean] and [corrected sample standard deviation][standard-deviation].
```

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md:
```diff
@@ -0,0 +1,234 @@
+<section class="notes">
+
+## Notes
+
```
```suggestion

- Input values are **not** type checked. If non-numeric inputs are possible, you are advised to type check and handle accordingly **before** passing the value to the accumulator function.
```

ShabiShett07 replied:
This specifies that non numeric inputs are allowed in this package

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/examples/index.js:
```diff
@@ -0,0 +1,68 @@
+* limitations under the License.
+*/
+
+'use strict';
```
I think the examples should be updated as we need to pass `NaN` values too in the input, please refer other similar packages.

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md:
````diff
@@ -0,0 +1,234 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var randu = require( '@stdlib/random/base/randu' );
````
Same here

ShabiShett07 commented:
@saad-imran-dev, I think your implementation is missing `repl.txt` file

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/lib/index.js:
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+/**
+* Compute a moving arithmetic mean and corrected sample standard deviation incrementally, ignoring NaN values.
```
```suggestion
* Compute a moving arithmetic mean and corrected sample standard deviation incrementally, ignoring `NaN` values.
```

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/docs/types/index.d.ts:
```diff
@@ -0,0 +1,94 @@
+import { ArrayLike } from '@stdlib/types/array';
+
+/**
+* If provided a value, the accumulator function returns an updated moving arithmetic mean and corrected sample standard deviation. If provided NaN or no value, the accumulator function returns the current moving arithmetic mean and corrected sample standard deviation.
```
```suggestion
* If provided a value, the accumulator function returns an updated moving arithmetic mean and corrected sample standard deviation. If provided `NaN` or no value, the accumulator function returns the current moving arithmetic mean and corrected sample standard deviation.
```

ShabiShett07 replied:
Here and elsewhere

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/benchmark/benchmark.js:
```diff
@@ -0,0 +1,69 @@
+
+var bench = require( '@stdlib/bench' );
+var randu = require( '@stdlib/random/base/randu' );
+var pkg = require( '@stdlib/stats/incr/nanmmeanstdev/package.json' ).name;
+var incrnanmmeanstdev = require( '@stdlib/stats/incr/nanmmeanstdev/lib' );
```
```suggestion
var incrnanmmeanstdev = require( './../lib' );
```

Planeshifter replied:
Use relative paths for `require` statements consistent with the base `mmeanstdev` benchmark.

```suggestion
var pkg = require( './../package.json' ).name;
var incrnanmmeanstdev = require( './../lib' );
```

ShabiShett07 commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/test/test.js:
```diff
@@ -0,0 +1,437 @@
+var sqrt = require( '@stdlib/math/base/special/sqrt' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var EPS = require( '@stdlib/constants/float64/eps' );
+var incrnanmmeanstdev = require( '@stdlib/stats/incr/nanmmeanstdev/lib' );
```
```suggestion
var incrnanmmeanstdev = require( './../lib' );
```

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/lib/index.js

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/test/test.js

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev committed:
Update lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md

Co-authored-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>
Signed-off-by: Muhmmad Saad <106260977+saad-imran-dev@users.noreply.github.com>

saad-imran-dev commented:
> @saad-imran-dev, I think your implementation is missing `repl.txt` file

@ShabiShett07 I get this error whenever I try to add repl.txt and commit. Not able to resolve this error.

```
/stdlib/lib/node_modules/@stdlib/_tools/repl-txt/rules/match-signature/lib/main.js:61
                                if ( params[i].multiple !== sigParams[i-skip].multiple ) {

```
```
{{alias}}()
    Returns an accumulator function which incrementally computes a moving
    arithmetic mean and corrected sample standard deviation, ignoring
    `NaN` values.

    The `W` parameter defines the number of values over which to compute the
    moving arithmetic mean and corrected sample standard deviation.

    If provided a value, the accumulator function returns updated accumulated
    values. If not provided a value, the accumulator function returns
    the current moving accumulated values.

    As `W` values are needed to fill the window buffer, the first `W-1` returned
    accumulated values are calculated from smaller sample sizes. Until the
    window is full, each returned accumulated value is calculated from all
    provided values.

    Parameters
    ----------
    out: Array|TypedArray (optional)
        Output array.

    W: integer
        Window size.

    Returns
    -------
    acc: Function
        Accumulator function.

    Examples
    --------
    > var accumulator = {{alias}}( 3 );
    > var v = accumulator()
    null
    > v = accumulator( 2.0 )
    [ 2.0, 0.0 ]
    > v = accumulator( -5.0 )
    [ -1.5, ~4.95 ]
    > v = accumulator( NaN )
    [ -1.5, ~4.95 ]
    > v = accumulator( 3.0 )
    [ 0.0, ~4.36 ]
    > v = accumulator( 5.0 )
    [ 1.0, ~5.29 ]
    > v = accumulator( NaN )
    [ 1.0, ~5.29 ]
    > v = accumulator()
    [ 1.0, ~5.29 ]

    See Also
    --------
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/docs/types/index.d.ts:
```diff
@@ -0,0 +1,94 @@
+* @example
+* var Float64Array = require( '@stdlib/array/float64' );
+*
+* var accumulator = incrmnanmeanstdev( new Float64Array( 2 ), 3 );
+*
+* var mm = accumulator();
+* // returns null
+*/
+declare function incrmnanmeanstdev( out: ArrayLike<number>, window: number ): accumulator;
```
The function name is `incrmnanmeanstdev` on lines 49, 54, 89, and 94, but the module exports `incrnanmmeanstdev`. The transposed letters will cause the TypeScript declarations to not match. Should be `incrnanmmeanstdev` to align with the module export and `test.ts`.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/lib/index.js:
Missing `docs/repl.txt`. See `@stdlib/stats/incr/mmeanstdev/docs/repl.txt` for the expected format.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/README.md:
```diff
@@ -0,0 +1,235 @@
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/stats/incr/mmeanstdev`][@stdlib/stats/incr/meanstdev]</span><span class="delimiter">: </span><span class="description">compute a moving arithmetic mean and corrected sample standard deviation incrementally.</span>
+-   <span class="package-name">[`@stdlib/stats/incr/meanstdev`][@stdlib/stats/incr/meanstdev]</span><span class="delimiter">: </span><span class="description">compute an arithmetic mean and corrected sample standard deviation incrementally.</span>
+-   <span class="package-name">[`@stdlib/stats/incr/mmean`][@stdlib/stats/incr/mmean]</span><span class="delimiter">: </span><span class="description">compute a moving arithmetic mean incrementally.</span>
+-   <span class="package-name">[`@stdlib/stats/incr/mmeanvar`][@stdlib/stats/incr/mmeanvar]</span><span class="delimiter">: </span><span class="description">compute a moving arithmetic mean and unbiased sample variance incrementally.</span>
+-   <span class="package-name">[`@stdlib/stats/incr/mstdev`][@stdlib/stats/incr/mstdev]</span><span class="delimiter">: </span><span class="description">compute a moving corrected sample standard deviation incrementally.</span>
+
+</section>
```
Broken link reference: display text says `mmeanstdev`, but the link anchor points to `[@stdlib/stats/incr/meanstdev]`. These are different packages.

Also, leave the "See Also" section empty. It is auto-populated during the release process.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/package.json:
```diff
@@ -0,0 +1,80 @@
+{
+  "name": "@stdlib/stats/incr/nanmmeanstdev",
+  "version": "0.0.0",
+  "description": "Compute a moving arithmetic mean and corrected sample standard deviation incrementally, ignoring NaN values",
```
Missing trailing period.

```suggestion
  "description": "Compute a moving arithmetic mean and corrected sample standard deviation incrementally, ignoring NaN values.",
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmmeanstdev/examples/index.js:
Missing `NaN` handling in examples. Use `bernoulli` to inject `NaN` values. See `stats/incr/nanmeanabs` examples for the pattern.

Planeshifter reviewed (changes requested):
Thanks for the PR! Left some feedback

Planeshifter committed:
fix: address review feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_kebab_case
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
docs: remove method section from delegating wrapper

Planeshifter committed:
style: align with base package conventions

Planeshifter committed:
docs: fix REPL doctest failures

Planeshifter committed:
docs: simplify REPL examples

Planeshifter reviewed (approved):
Let's get it in!
